modules/ads_lib/src/main/java/com/google/api/ads/adwords/lib/utils/logging/ReportServiceLogger.java

```diff
@@ -0,0 +1,191 @@
+import java.util.Map.Entry;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+
+/**
+ * Logger that logs report requests and responses according to the following rules.<br>
+ *
+ * <ul>
+ *   <li>Log successful requests (header and payload) and responses to INFO.
+ *   <li>Log failed requests (header and payload) and responses to WARN.
+ * </ul>
+ */
+public class ReportServiceLogger {
+
+  /**
+   * Headers whose value should not be logged because they contain sensitive information. Entries
+   * are in lowercase, but scrubbing should be case-insensitive.
+   */
+  @VisibleForTesting
+  static final Set<String> SCRUBBED_HEADERS =
+      Sets.newHashSet("authorization", "authtoken", "password", "developertoken");
+  /** Value to log for a scrubbed header. */
+  @VisibleForTesting static final String REDACTED_HEADER = "REDACTED";
+  /** String to log in place of the report response contents. */
+  @VisibleForTesting static final String REDACTED_REPORT_DATA = "REDACTED REPORT DATA";
+
+  private static final String CLIENT_CUSTOMER_ID = "clientCustomerId";
+
+  private final RemoteCallLoggerDelegate loggerDelegate;
+
+  /**
+   * Constructor that takes an injected logger identified by name.
+   *
+   * @param reportLogger underlying SLF4J logger for report service interactions
+   */
+  @Inject
+  private ReportServiceLogger(@Named(AdWordsLoggingModule.REPORT_LOGGER_NAME) Logger reportLogger) {
+    this(new RemoteCallLoggerDelegate(reportLogger, reportLogger, null, RemoteCallType.HTTP));
+  }
+
+  /** @param remoteCallLoggerDelegate */
+  ReportServiceLogger(RemoteCallLoggerDelegate remoteCallLoggerDelegate) {
+    this.loggerDelegate = remoteCallLoggerDelegate;
+  }
+
+  /**
+   * Logs the specified request and response information.
+   *
+   * <p>Note that in order to avoid any temptation to consume the contents of the response, this
+   * does <em>not</em> take an {@link com.google.api.client.http.HttpResponse} object, but instead
+   * accepts the status code and message from the response.
+   */
+  public void logRequest(
+      @Nullable HttpRequest request, int statusCode, @Nullable String statusMessage) {
+    boolean isSuccess = HttpStatusCodes.isSuccess(statusCode);
+    if (!loggerDelegate.isSummaryLoggable(isSuccess)
+        && !loggerDelegate.isDetailsLoggable(isSuccess)) {
+      return;
+    }
+
+    // Populate the RequestInfo builder from the request.
+    RequestInfo requestInfo = buildRequestInfo(request);
+
+    // Populate the ResponseInfo builder from the response.
+    ResponseInfo responseInfo = buildResponseInfo(request, statusCode, statusMessage);
+
+    RemoteCallReturn.Builder remoteCallReturnBuilder =
+        new RemoteCallReturn.Builder().withRequestInfo(requestInfo).withResponseInfo(responseInfo);
+    if (!isSuccess) {
+      remoteCallReturnBuilder.withException(
+          new ReportException(String.format("%s: %s", statusCode, statusMessage)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: statusCode (declared at line 91). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 108. Consider formatting this data with &#39;%d&#39; instead.